config.ru

```diff
@@ -1,4 +1,13 @@
 # This file is used by Rack-based servers to start the application.
 
+if ENV.fetch("KILL_UNICORNS", false) && (Rails.env.production? || Rails.env.staging?)
+  # Gracefully restart individual unicorn workers if they have:
+  # - performed between 10000 and 15000 requests
+  # - grown in memory usage to between 500 and 600 MB
+  require 'unicorn/worker_killer'
+  use Unicorn::WorkerKiller::MaxRequests, 10_000, 15_000
+  use Unicorn::WorkerKiller::Oom, (500 * (1024**2)), (600 * (1024**2))
```

Comment:
how many requests do we have per day? We should base this config on that number too avoid too many restarts.

I wonder if the mem values are not too low and will trigger too many restarts. On your screenshot the other two processes are taking 700MB and 555MB right? Maybe 1GB would be better?

Also, what's the upper limit for? if the worker is taking more than the max 600MB will it not be restarted?

Reply:
after this is live, is there a way to monitor how many restarts were triggered by these rules?

Reply:
The actual limit is decided by rand() between memory_limit_min and memory_limit_max per worker, to prevent all workers to be dead at the same time.

Reply:
> to prevent all workers to be dead at the same time

Exactly. There should only be one restarted at a time, so the others are still available. And it only does the restart whilst a worker is _between_ requests, not busy.

> how many requests do we have per day?

I took a quick look at UK Prod and got ~64k for today. So I guess with high usage, it would be roughly every 2 hours. Maybe we can go higher for both of these settings...

Reply:
I've updated the values. Also note this will not be used unless the instance explicitly enables it by setting `ENV['KILL_UNICORNS']`.

Reply:
Some interesting stats form French Prod for comparison:

![Screenshot from 2020-04-04 22-47-25](https://user-images.githubusercontent.com/9029026/78461064-72cf1e80-76c6-11ea-9848-2d559c5f1718.png)

Here we can see:
- the workers have all been running for ~256 hours
- two of them are sitting nicely at 590 and 620 MB, which looks reasonable
- two of them are eating 1.5 and 1.7 GB, which really doesn't look great

I think if we had `unicorn-worker-killer` running, French Prod would have roughly an extra 2GB of free memory right now. Likwise, UK Prod wouldn't be deep in swap today.

Reply:
These numbers from `ps` are actually in kilobytes, so they're fractionally lower than advertised: 
1711476 / 1024 = **1671 MB**
1494168 / 1024 = **1459 MB**
620824 / 1024 = **606 MB**
594400 / 1024 = **580 MB**

I think if a worker goes over 700MB, it's time for a shift change.

Maybe we can add some ENV vars to make these upper and lower bounds configurable though...

Reply:
More stats; last night on UK Prod we went up to 16GB of RAM. This morning our four workers look like this:

![Screenshot from 2020-04-05 09-50-39](https://user-images.githubusercontent.com/9029026/78469608-2665fc00-7723-11ea-8166-45a917dcfb99.png)

That's 3 sitting nicely at 470-520 MB, and one at **~6GB**...

Reply:
And finally, some terrifying stats from Aus. This is the memory graph for the last 2 days (it was fairly flat earlier in the week):

![Screenshot from 2020-04-05 10-11-58](https://user-images.githubusercontent.com/9029026/78470006-3af7c380-7726-11ea-98a4-23b9c70c431c.png)

And the 3 unicorn workers currently:

![Screenshot from 2020-04-05 10-12-56](https://user-images.githubusercontent.com/9029026/78470022-58c52880-7726-11ea-9db2-3ee9f4a77ee1.png)

That's 3 workers using between **3.1 and 3.4 GB** each...

Reply:
@mkllnk